Allow `/` in relative paths for `using module` (#7424)

jcotton42 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->

Per the title, this makes `using module` work with a relative path that uses forward slashes. See [the issue this is closing](https://github.com/PowerShell/PowerShell/issues/7424) for more details.

Fix #7424.

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

jcotton42 committed:
Close #7424. Allow / in relative paths for `using module`

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12492) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12492) before we can accept your contribution.<br/><br/>:x: jcotton42 [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12492)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=12492) it.</sub>

vexx32 commented:
@PoshChan please retry macos

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS`

vexx32 reviewed (approved):
Looks good to me, nice work! 😊 💖

iSazonov commented on test/powershell/Language/Parser/UsingModule.Tests.ps1:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+Describe "Using module" -Tag "CI" {
```
We have scripting.Classes.using.tests.ps1 for such tests. Please move the tests to the file.

iSazonov commented on test/powershell/Language/Parser/UsingModule.Tests.ps1:
```diff
@@ -0,0 +1,23 @@
+Describe "Using module" -Tag "CI" {
+    BeforeAll {
+        Push-Location $PSScriptRoot
+        New-Item tmp -ItemType Directory
```
Can we use Pester TestDrive?

jcotton42 replied:
I was unaware of that Pester feature, I will see if it will work here.

jcotton42 committed:
Move test to correct test file

jcotton42 commented:
Moved the test to the right file and updated it to use Pester's TestDrive

jcotton42 commented on test/powershell/Language/Classes/scripting.Classes.using.tests.ps1:
```diff
@@ -339,7 +339,7 @@ using module Foo
     }
 
-        Context 'Side by side' {
+    Context 'Side by side' {
```
Happened to notice the indentation of this line was wrong, decided to fix it since I was already in here

jcotton42 commented:
CI for all 3 OSes failed, but it's code that, as far as I can tell, was unaffected by my change. And the tests ran just fine on the last commit, and the only change between the last commit and the current one is I moved my test and updated it.

vexx32 commented:
@PoshChan please retry all

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

iSazonov commented on src/System.Management.Automation/engine/parser/SymbolResolver.cs:
```diff
@@ -498,8 +498,8 @@ private Collection<PSModuleInfo> GetModulesFromUsingModule(UsingStatementAst usi
-                // case 1: relative path. Relative for file in the same folder should include .\
-                bool isPath = fullyQualifiedNameStr.Contains(@"\");
+                // case 1: relative path. Relative for file in the same folder should include .\ or ./
+                bool isPath = fullyQualifiedNameStr.Contains(@"\") || fullyQualifiedNameStr.Contains("/");
```
```suggestion
                bool isPath = fullyQualifiedNameStr.Contains('\\') || fullyQualifiedNameStr.Contains('/');
```

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.using.tests.ps1:
```diff
@@ -523,6 +523,15 @@ using module FooForPaths
+
+            "using module ./$name.psm1; Get-Foo" | Set-Content "TestDrive:\modules\$name.ps1"
+
+            { & "TestDrive:\modules\$name.ps1" } | Should -Not -Throw
```
Please use _positive_ test (check Get-Foo return).
Also please replace `Foo` with value name like Get-TestString.

jcotton42 committed:
Resolve feedback

Contains char instead of a string, and changing how the test checked for
success

jcotton42 committed:
Add regression test to ensure .\ still works for using module

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.using.tests.ps1:
```diff
@@ -532,6 +532,15 @@ using module FooForPaths
+
+            "using module .\$name.psm1; Get-TestString" | Set-Content "TestDrive:\modules\$name.ps1"
+
+            & "TestDrive:\modules\$name.ps1" | Should -EQ "Worked"
```
```suggestion
            & "TestDrive:\modules\$name.ps1" | Should -BeExactly "Worked"
```

Above too.

jcotton42 committed:
Use BeExactly in Pester tests

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.using.tests.ps1:
```diff
@@ -523,6 +523,24 @@ using module FooForPaths
+            & "TestDrive:\modules\$name.ps1" | Should -BeExactly "Worked"
+        }
+
+        It 'can be accessed by relative path with .\' {
```
Both new tests use identical code and we should merge it in one with using `-TestCases`.

jcotton42 replied:
I can do this as soon as I get off work (about 5 PM EDT), best to change this to have best practices since it's such a simple change.

PoshChan commented:
@TravisEz13, this is the reminder you requested 4 hours ago

jcotton42 committed:
Refactor tests to use TestCases

jcotton42 commented:
`TestCase` refactor complete, should be ready for merging once CI passes.

jcotton42 commented:
@PoshChan please retry ssh

PoshChan commented:
@jcotton42, you are not authorized to request a rebuild

vexx32 commented:
@PoshChan please retry ssh

PoshChan commented:
@vexx32, I do not understand the build target(s) `ssh`; I only allow `static`, `windows`, `macos`, `linux`, `all`

vexx32 commented:
@PoshChan please retry all

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.using.tests.ps1:
```diff
@@ -524,20 +524,12 @@ using module FooForPaths
 
-        It 'can be accessed by relative path with ./' {
-            $name = 'relative-forward-slash-paths'
+        It 'can be accessed by relative path with .<Separator>' -TestCases @{Separator = '\'}, @{Separator = '/'} {
```
```suggestion
        It 'can be accessed by relative path with .<Separator>' -TestCases @(
            @{ Separator = '\' },
            @{ Separator = '/' }
        ) {
```

jcotton42 committed:
Update TestCases formatting

Co-authored-by: Ilya <darpa@yandex.ru>

PoshChan commented:
@TravisEz13, this is the reminder you requested 1 hour ago

iSazonov commented:
@jcotton42 Thanks for your contribution!
